signac/core/collection_json.py

```diff
@@ -27,6 +90,7 @@ def __init__(self, filename=None, write_concern=False, **kwargs):
         self._write_concern = write_concern
         kwargs['name'] = filename
         super().__init__(**kwargs)
+        self._validators.append(JSONFormatValidator)
```

Comment:
@vishav1771 as per my comment on the docstring of the `JSONFormatValidator` class, that format validator is not specific to just this collection but could be used for any collection. This also means that `SyncedCollection` should expose an API to register validators to specific subclasses (like `JSONCollection`). However, I don't think that the validation should be controlled by our infrastructure, i.e. a user of the `SyncedCollection` family of classes should be able to freely mix and match validators and backends. I think the best way to enable this is for `SyncedCollection` to define a method `add_validator` or accept validators as a constructor argument. So the prototype in signac I'm imagining would look something like

```
# Need to import from wherever these actually live.
from synced_collection... import JSONDict, JSONFormatValidator

class Job:
    def __init__(self, ...):
        ...
        # Option 1
        self._sp = JSONDict(..., validator=JSONFormatValidator)
        # Option 2
        self._sp = JSONDict(...)
        self._sp.add_validator(JSONFormatValidator)

```

An alternative would be to do this on a class level, i.e. to define `add_validator` as a `classmethod` of `SyncedCollection`. This approach introduces complexities such as whether you should inherit validators from parent classes. For instance, if you do `JSONCollection.add_validator(...)` and `JSONDict.add_validator(...)`, I would probably expect `JSONDict` to use the validators added to `JSONCollection` irrespective of the order in which those two calls occurred. On the plus side, this approach is more concise since you don't have to reassign the validator to every instance. @csadorf thoughts?

Reply:
I need to walk back one part of this: certain backends may _require_ some validation, in which case those validators should be applied automatically. In particular, all `JSONCollection` instances need to have the `JSONFormatValidator` applied (even outside the signac context), whereas in signac we plan to apply that validator irrespective of backend because we require the data to be valid JSON no matter what backend is used. So you probably still need to add the validator here.

Since there need to be some default validators for certain classes, that might lead us to prefer the `classmethod` approach I described, but it also wouldn't be hard to implement either of the instance-based approaches and simply have a separate class constant that just stores the default validators used for instances of a class. As a result, I think both approaches are still feasible even with this requirement.